File: Python/ExampleTests/json_glue.py
from common import *
from production import SimpleJson


class JsonGlue:
    DNC_STRING = '?DNC?'

    def __init__(self):
        self.simple_class_values = []
        self.given_json = ''
        self.actual_json = ''
        self.parsed_object = []

    # SimpleJson takes plain name/value maps, so it stays independent of the
    # generated test classes. These two moves are the whole of the mapping —
    # the conversion itself belongs to SimpleJson.
    @staticmethod
    def _fields_of(value: SimpleClassString) -> dict:
        return {'anInt': value.an_int, 'aString': value.a_string}

    @staticmethod
    def _object_of(fields: dict) -> SimpleClassString:
        return SimpleClassString(fields.get('anInt'), fields.get('aString'))

    def given_one_object_is(self, values: list):
        for value in values:
            print(value)
        self.simple_class_values = values
        self.actual_json = SimpleJson.to_object(self._fields_of(values[0]))

    def then_json_should_be(self, value: str):
        print(value)
        # Text to text, with the whitespace between tokens removed from both
        # sides. Whitespace inside a quoted value is kept.
        assert SimpleJson.without_whitespace(value) == \
               SimpleJson.without_whitespace(self.actual_json)

    def given_json_is(self, value: str):
        print(value)
        self.given_json = value
        self.parsed_object = [self._object_of(SimpleJson.parse_object(value))]

    def then_the_converted_object_is(self, values: list):
        for value in values:
            print(value)
        assert values == self.parsed_object

    def given_a_table_is(self, values: list):
        for value in values:
            print(value)
        self.simple_class_values = values
        self.actual_json = SimpleJson.to_array([self._fields_of(v) for v in values])

    def then_json_for_table_should_be(self, value: str):
        print(value)
        assert SimpleJson.without_whitespace(value) == \
               SimpleJson.without_whitespace(self.actual_json)

    def given_json_for_table_is(self, value: str):
        print(value)
        self.given_json = value
        self.parsed_object = [self._object_of(row)
                              for row in SimpleJson.parse_array(value)]

    def then_the_converted_table_should_be(self, values: list):
        for value in values:
            print(value)
        assert values == self.parsed_object

File: JavaScript/ExampleTests/json_glue.js
import { SimpleClassString } from "./common/index.js";
import { toObject, toArray, parseObject, parseArray, withoutWhitespace }
  from "./production/SimpleJson.js";

export class JsonGlue {
  static DNC_STRING = "?DNC?";

  constructor() {
    this.simpleClassValues = [];
    this.givenJson = "";
    this.actualJson = "";
    this.parsedObject = [];
  }

  // SimpleJson takes plain name/value pairs, so it stays independent of the
  // generated test classes. These two moves are the whole of the mapping — the
  // conversion itself belongs to SimpleJson.
  static #fieldsOf(value) {
    return [["anInt", value.anInt], ["aString", value.aString]];
  }

  static #objectOf(fields) {
    const map = new Map(fields);
    return new SimpleClassString(map.get("anInt"), map.get("aString"));
  }

  givenOneObjectIs(values) {
    values.forEach(v => console.log(v.toString()));
    this.simpleClassValues = values;
    this.actualJson = toObject(JsonGlue.#fieldsOf(values[0]));
  }

  thenJsonShouldBe(value) {
    console.log(value);
    // Text to text, with the whitespace between tokens removed from both sides.
    // Whitespace inside a quoted value is kept.
    expect(withoutWhitespace(value)).toBe(withoutWhitespace(this.actualJson));
  }

  givenJsonIs(value) {
    console.log(value);
    this.givenJson = value;
    this.parsedObject = [JsonGlue.#objectOf(parseObject(value))];
  }

  thenTheConvertedObjectIs(values) {
    values.forEach(v => console.log(v.toString()));
    expect(SimpleClassString.equalLists(values, this.parsedObject)).toBe(true);
  }

  givenATableIs(values) {
    values.forEach(v => console.log(v.toString()));
    this.simpleClassValues = values;
    this.actualJson = toArray(values.map(v => JsonGlue.#fieldsOf(v)));
  }

  thenJsonForTableShouldBe(value) {
    console.log(value);
    expect(withoutWhitespace(value)).toBe(withoutWhitespace(this.actualJson));
  }

  givenJsonForTableIs(value) {
    console.log(value);
    this.givenJson = value;
    this.parsedObject = parseArray(value).map(row => JsonGlue.#objectOf(row));
  }

  thenTheConvertedTableShouldBe(values) {
    values.forEach(v => console.log(v.toString()));
    expect(SimpleClassString.equalLists(values, this.parsedObject)).toBe(true);
  }
}
